src/gui/layout/qgslayoutattributeselectiondialog.cpp

```diff
@@ -211,57 +220,57 @@ QVariant QgsLayoutAttributeTableColumnModel::headerData( int section, Qt::Orient
   }
 }
 
-bool QgsLayoutAttributeTableColumnModel::setData( const QModelIndex &index, const QVariant &value, int role )
+bool QgsLayoutAttributeTableColumnModelBase::setData( const QModelIndex &index, const QVariant &value, int role )
 {
   if ( !index.isValid() || role != Qt::EditRole || !mTable )
-  {
     return false;
-  }
-  if ( index.row() >= mTable->columns().length() )
-  {
+
+  if ( index.row() >= columns().length() )
     return false;
-  }
 
-  //get column for index
-  QgsLayoutTableColumn *column = columnFromIndex( index );
-  if ( !column )
-  {
+  if ( index.column() > displayedColumns().count() )
     return false;
-  }
 
-  switch ( index.column() )
+  Column col = displayedColumns()[index.column()];
+  switch ( col )
   {
-    case 0:
+    case Attribute:
       // also update column's heading, if it hasn't been customized
-      if ( column->heading().isEmpty() || ( column->heading() == column->attribute() ) )
+      if ( columns()[index.row()].heading().isEmpty() || ( columns()[index.row()].heading() == columns()[index.row()].attribute() ) )
       {
-        column->setHeading( value.toString() );
+        columns()[index.row()].setHeading( value.toString() );
         emit dataChanged( createIndex( index.row(), 1 ), createIndex( index.row(), 1 ) );
       }
-      column->setAttribute( value.toString() );
+      columns()[index.row()].setAttribute( value.toString() );
       emit dataChanged( index, index );
       return true;
-    case 1:
-      column->setHeading( value.toString() );
+
+    case Heading:
+      columns()[index.row()].setHeading( value.toString() );
       emit dataChanged( index, index );
       return true;
-    case 2:
-      column->setHAlignment( Qt::AlignmentFlag( value.toInt() & Qt::AlignHorizontal_Mask ) );
-      column->setVAlignment( Qt::AlignmentFlag( value.toInt() & Qt::AlignVertical_Mask ) );
+
+    case Alignment:
+      columns()[index.row()].setHAlignment( Qt::AlignmentFlag( value.toInt() & Qt::AlignHorizontal_Mask ) );
```

Comment:
Lots of calls to `columns()[index.row()]` here -- could this be called once and stored for reuse in this function?

Reply:
sadly not as we are updating a column. But I replaced `index.row()` by `r`, makes it easier to read.

Reply:
hum, thanks for the free introduction course ;)